src/app/outbound.rs

```diff
@@ -302,13 +205,8 @@ pub mod orig_proto_upgrade {
         type Error = M::Error;
 
         fn make(&self, endpoint: &Endpoint) -> Result<Self::Value, Self::Error> {
-            if endpoint.can_use_orig_proto()
-                && !endpoint.dst.settings.is_http2()
-                && !endpoint.dst.settings.is_h1_upgrade()
-            {
-                let mut upgraded = endpoint.clone();
-                upgraded.dst.settings = Settings::Http2;
-                self.inner.make(&upgraded).map(|i| svc::Either::A(i.into()))
+            if endpoint.can_use_orig_proto() {
```

Comment:
Well, I've done some reading, and don't quite get it yet. It seems *something* caches the `make` call on the client layer. I checked out this branch locally, and put together a test that ensured an orig-proto connection was setup first, and then tried an upgrade request, and it properly knew to skip the cache. :shrug:

Reply:
This is supposed to work as follows:

- the stack first binds via`settings::router` which produces one of 5 clients (http2, http1/absolute-form/is-upgrade) https://github.com/linkerd/linkerd2-proxy/blob/master/src/proxy/http/settings.rs#L9-L29
- then, within each set of settings, the orig-proto module does an in-line check to determine whether the version should be changed

Reply:
I looked into this again because I was worried I missed something but I think it makes sense to me now... My last explanation isn't quite right:

https://github.com/linkerd/linkerd2-proxy/blob/19c71d2eb47772dbf55433e7d23326c3e9c9ffb8/src/app/main.rs#L311-L319

In the endpoint stack, we check whether the endpoint supports upgrading, and if it does we apply the orig_proto:Upgrade middleware. The middleware flips request versions as appropriate. Then, the settings::router module routes the request to the appropriate client based on the request's Version:

https://github.com/linkerd/linkerd2-proxy/blob/19c71d2eb47772dbf55433e7d23326c3e9c9ffb8/src/proxy/http/settings.rs#L30-L51

@seanmonstar does this sound right to you?

Reply:
Yep!